crates/shared/src/price_estimation/native.rs

```diff
@@ -48,34 +39,60 @@ impl NativePriceEstimator {
             price_estimation_amount,
         }
     }
+
+    fn query(&self, token: &H160) -> Query {
+        Query {
+            sell_token: *token,
+            buy_token: self.native_token,
+            in_amount: self.price_estimation_amount,
+            kind: OrderKind::Buy,
+        }
+    }
 }
 
 #[async_trait::async_trait]
 impl NativePriceEstimating for NativePriceEstimator {
-    async fn estimate_native_prices(
-        &self,
-        tokens: &[H160],
-    ) -> Vec<Result<f64, PriceEstimationError>> {
-        let native_token_queries: Vec<_> = tokens
-            .iter()
-            .map(|token| Query {
-                sell_token: *token,
-                buy_token: self.native_token,
-                in_amount: self.price_estimation_amount,
-                kind: OrderKind::Buy,
-            })
-            .collect();
+    fn estimate_native_prices<'a>(
+        &'a self,
+        tokens: &'a [H160],
+    ) -> BoxStream<'_, (usize, NativePriceEstimateResult)> {
+        let stream = async_stream::stream!({
+            let queries: Vec<_> = tokens.iter().map(|token| self.query(token)).collect();
+            let mut inner = self.inner.estimates(&queries);
+            while let Some((i, result)) = inner.next().await {
+                let result = result.map(|estimate| estimate.price_in_buy_token_f64(&queries[i]));
+                yield (i, result)
+            }
+        });
+        stream.boxed()
+    }
+}
 
-        let estimates = vec_estimates(self.inner.as_ref(), &native_token_queries).await;
+pub async fn native_single_estimate(
+    estimator: &dyn NativePriceEstimating,
+    token: &H160,
+) -> NativePriceEstimateResult {
+    estimator
+        .estimate_native_prices(std::slice::from_ref(token))
+        .next()
+        .await
+        .unwrap()
+        .1
+}
 
-        estimates
-            .into_iter()
-            .zip(native_token_queries.iter())
-            .map(|(estimate, query)| {
-                estimate.map(|estimate| estimate.price_in_buy_token_f64(query))
-            })
-            .collect()
+pub async fn native_vec_estimates(
+    estimator: &dyn NativePriceEstimating,
+    queries: &[H160],
+) -> Vec<NativePriceEstimateResult> {
+    let mut results = vec![None; queries.len()];
+    let mut stream = estimator.estimate_native_prices(queries);
+    while let Some((index, result)) = stream.next().await {
+        results[index] = Some(result);
     }
+    let results = results.into_iter().flatten().collect::<Vec<_>>();
+    // Check that every query has a result.
+    debug_assert_eq!(results.len(), queries.len());
+    results
```

Comment:
The reason I have these as separate functions and not part of the trait is that moving them into the trait makes the overridable which I don't want. Agree it is slightly more annoying to use but also shows intent slightly more.
Your implementation makes sense too. Initially I used this one so this debug assert would tell me if a price estimator had a bug where it would not use an index or use an index multiple times. Soonish most uses of these helpers should be gone gone anyway. I also realize now I can make this function generic so that it works with both price estimators.

Reply:
One option for the helper functions is to shove them in an extension trait. Its just syntactic sugar though - so I'm indifferent.

> An alternative implementation could be

Nice suggestion - its concise.